text/0000-track-unique-history-location-state.md

```diff
@@ -0,0 +1,93 @@
+- Start Date: 2016/12/05
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Track unique history location states
+
+# Motivation
+
+The path alone does not provide enough information. For example, if you
+visit page A, scroll down, then click on a link to page B, then click on
+a link back to page A. Your actual browser history stack is [A, B, A].
+Each of those nodes in the history should have their own unique scroll
+position. In order to record this position we need an ID that is unique
+for each node in the history.
+
+# Detailed design
+
+Code: [PR#14011](https://github.com/emberjs/ember.js/pull/14011)
+
+We simply add a `stateCounter` so we can track uniqueness on two
+dimesions. Both `path` and the generated `id`.
+
+This API will allow other libraries to reflect upon each location to
+determine unique state. For example,
+[ember-router-scroll](https://github.com/dollarshaveclub/ember-router-scroll)
+is making use of a [modified `Ember.HistoryLocation` object to get this
+behavior](https://github.com/dollarshaveclub/ember-router-scroll/blob/master/addon/locations/router-scroll.js).
+
+Tracking unique state is required when setting the scroll position
+properly based upon where you are in the history stack, as described in
+[Motivation](#motivation)
+
+# How We Teach This
+
+We could describe what meta data is generated for each location in the
+history stack. For example, it could look like:
+
+```js
+// visit page A
+
+[
+  { path: '/', id: 1 }
+]
+
+// visit page B
+
+[
+  { path: '/about', id: 2 },
+  { path: '/', id: 1 }
+]
+
+// visit page A
+
+[
+  { path: '/', id: 3 },
+  { path: '/about', id: 2 },
+  { path: '/', id: 1 }
+]
+
+// click back button
+
+[
+  { path: '/about', id: 2 },
+  { path: '/', id: 1 }
+]
+```
+
+# Drawbacks
+
+None that I'm aware of
```

Comment:
@krisselden reading https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage Its not clear to me on how to track unique page visits. If I visit A, then B, then A with a history stack of [A, B, A] how would sessionStorage API be used? Would the page be the key? If yes, then that only have a single dimension of uniqueness and doesn't account for multiple page visits. (which this RFC addresses)

If the entire history object is to be stored in sessionState, is this something that should be restored at app launch? That would seem like an odd user experience.